### DISCUSSION TITLE
Client redirect_uri check must be compared using exact string matching?

### DISCUSSION THREAD

**Discussion by cgeorgilakis** [keycloak/keycloak#25002] [Ideas]
From Keycloak 21.1.2 as described in [this issue](https://github.com/keycloak/keycloak/issues/24792), Keycloak has a bug with redirect_uri with uppercase letters.
This was fixed in version 23.0.0 in a bad and insecure way with [this PR](https://github.com/keycloak/keycloak/pull/24819).

Maybe OAuth 2.0 allows case-insensitive redirect_uri check. Could you provide the specification part of it?
For sure, [OAuth 2.0 Security Best Current Practice](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#name-protecting-redirect-based-f) mentions :

> When comparing client redirect URIs against pre-registered URIs, authorization servers MUST utilize exact string matching except for port numbers in localhost redirection URIs of native apps, see [Section 4.1.3](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#iuv_countermeasures). We believe Keycloak should follow OAuth 2.0 Security Best Current Practice.

Moreover, this will be a step for implementing [OAuth 2.1](https://oauth.net/2.1/). :

> The major differences from OAuth 2.0 are listed below.
> ...
> Redirect URIs must be compared using exact string matching

We believe that Keycloak should has 'redirect_uri check must be compared using exact string matching' from now. What is your opinion?
I have created a [task issue](https://github.com/keycloak/keycloak/issues/25001) also with this. I did not the suitable issue type.

**Comment by rmartinc** [comment]
Hi @cgeorgilakis!

Previously keycloak was already using lowercase for hostname in `http`. This was extended with the PR to any protocol as it was doing the lowercase incorrectly for other protocols (`https` or custom ones). I also considered about removing any lowercase at all and simplify things. So personally I'm open to do it but this will create issues in upgrades for current `http` checks (which was already doing the lowercase for the passed redirect).

@mposolda @pedroigor WDYT?

**Comment by cgeorgilakis** [comment]
We use version 18 and we want to migrate to version 22.
In version 18 exact match is working with https at least ( I suppose for all versions before [a1cfe6e](https://github.com/keycloak/keycloak/commit/a1cfe6e24e5b34792699a00b8b4a8016a5929e3a) ).

Version 22 redirect_uri check is a problem for us. It breaks previous behaviour and it is against OAuth 2.0 Security Best Current Practice.
So, we find your fix and we want to open a discussion about this.
Unfortunatelly, the same happens for version 23 that includes the mentioned PR.

@mposolda @pedroigor WDYT?

**Reply by pedroigor** [reply]
We should make things simpler and follow the recommendation from specs.
